pip/vcs/git.py

```diff
@@ -74,27 +74,34 @@ def export(self, location):
                 show_stdout=False, cwd=temp_dir.path
             )
 
-    def check_rev_options(self, rev, dest, rev_options):
-        """Check the revision options before checkout to compensate that tags
-        and branches may need origin/ as a prefix.
-        Returns the SHA1 of the branch or tag if found.
+    def get_ref_sha(self, dest, ref):
         """
-        revisions = self.get_short_refs(dest)
-
-        origin_rev = 'origin/%s' % rev
-        if origin_rev in revisions:
-            # remote branch
-            return [revisions[origin_rev]]
-        elif rev in revisions:
-            # a local tag or branch name
-            return [revisions[rev]]
-        else:
+        Return a commit hash for the reference if the given name is a ref.
+        Otherwise, return None.
+
+        Args:
+          dest: the repository directory.
+          ref: the ref name.
+        """
+        output = self.run_command(['show-ref', '--hash', ref], cwd=dest,
+                                  show_stdout=False, on_returncode='ignore')
+        # The git-show-ref command can return multiple references, so we
+        # return only the first.
+        for sha in output.strip().splitlines():
+            return sha
```

Comment:
By the way, can you think of any circumstances in which we should ever have to look at `refs/heads/`? Since we are always getting updates from the remote by first cloning or fetching, it seems possible that looking at `refs/remotes/origin/` or `refs/tags/` would always suffice.